python/pywatchman/aioclient.py

```diff
@@ -0,0 +1,340 @@
+    loop = asyncio.get_event_loop()
+    return await loop.run_in_executor(None, _resolve_sockname_helper)
+
+
+class AsyncTransport(object):
+    '''Communication transport to the Watchman Service.'''
+
+    async def activate(self, **kwargs):
+        '''Make the transport ready for use. Optional for subclasses.'''
+        pass
+
+    async def read(self, size):
+        '''Read 'size' bytes from the transport.'''
+        raise NotImplementedError()
+
+    async def write(self, buf):
+        '''Write 'buf' bytes to the transport.'''
+        raise NotImplementedError()
+
+    def close(self):
+        '''Close the transport. Optional for subclasses.'''
+        pass
+
+
+class AsyncUnixSocketTransport(AsyncTransport):
+    '''Local Unix domain socket transport supporting asyncio.'''
+
+    def __init__(self):
+        self.sockname = None
+        self.reader = None
+        self.writer = None
+
+    async def activate(self, **kwargs):
+        # Requires keyword-argument 'sockname'
+        reader, writer = await asyncio.open_unix_connection(kwargs['sockname'])
+        self.reader = reader
+        self.writer = writer
+
+    async def write(self, data):
+        self.writer.write(data)
+        await self.writer.drain()
+
+    async def read(self, size):
+        res = await self.reader.read(size)
+        return res
+
+    def close(self):
+        if self.writer:
+            self.writer.close()
+
+
+class AsyncCodec(object):
+    '''Communication encoding for the Watchman service.'''
+
+    def __init__(self, transport):
+        self.transport = transport
+
+    async def receive(self):
+        '''Read from the underlying transport, parse and return the message.'''
+        raise NotImplementedError()
+
+    async def send(self, *args):
+        '''Send the given message via the underlying transport.'''
+        raise NotImplementedError()
+
+    def close(self):
+        '''Close the underlying transport.'''
+        self.transport.close()
+
+
+# TODO(dhruvsinghal): Replace this with BSER v2 when its ready.
+class AsyncBserCodec(AsyncCodec):
+    '''Use the BSER encoding.'''
+
+    async def receive(self):
+        buf = [await self.transport.read(SNIFF_LEN)]
+        if not buf[0]:
+            raise WatchmanError('empty watchman response')
+
+        _1, _2, elen = bser.pdu_info(buf[0])
+
+        rlen = len(buf[0])
+        while elen > rlen:
+            buf.append(await self.transport.read(elen - rlen))
+            rlen += len(buf[-1])
+
+        response = b''.join(buf)
+        try:
+            res = self._loads(response)
+            return res
+        except ValueError as e:
+            raise WatchmanError('watchman response decode error: %s' % e)
+
+    async def send(self, *args):
+        cmd = bser.dumps(*args) # Defaults to BSER v1
+        await self.transport.write(cmd)
+
+    def _loads(self, response):
+        ''' Parse the BSER packet '''
+        return bser.loads(response) # Defaults to BSER v1
```

Comment:
I think we could TODO an improvement here, but at the moment it won't really be possible to release the GIL in the middle of the `loads` -- the BSER deserializer tends to create Python objects on the fly.